{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "os.environ[\"LANGCHAIN_TRACING_V2\"] = \"true\"\n",
    "os.environ[\"LANGCHAIN_API_KEY\"] = getpass.getpass()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import getpass\n",
    "import os\n",
    "\n",
    "os.environ[\"AZURE_OPENAI_API_KEY\"] = getpass.getpass()\n",
    "\n",
    "from langchain_openai import AzureChatOpenAI\n",
    "os.environ[\"AZURE_OPENAI_ENDPOINT\"] = \"https://gpt4edu-ncus.openai.azure.com/\"\n",
    "os.environ[\"AZURE_OPENAI_DEPLOYMENT_NAME\"] = \"GPT4o\"\n",
    "os.environ[\"AZURE_OPENAI_API_VERSION\"] = \"2023-12-01-preview\"\n",
    "llm = AzureChatOpenAI(\n",
    "    azure_endpoint=os.environ[\"AZURE_OPENAI_ENDPOINT\"],\n",
    "    azure_deployment=os.environ[\"AZURE_OPENAI_DEPLOYMENT_NAME\"],\n",
    "    openai_api_version=os.environ[\"AZURE_OPENAI_API_VERSION\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import getpass\n",
    "import os\n",
    "\n",
    "os.environ[\"Ir_API_KEY\"] = getpass.getpass()\n",
    "\n",
    "# from langchain_openai import ChatOpenAI\n",
    "# llm = ChatOpenAI(\n",
    "#     openai_api_base=\"https://api.wlai.vip/v1\",\n",
    "#     openai_api_key=os.environ[\"Ir_API_KEY\"],\n",
    "# )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'The article describes a study using cryo-electron microscopy (cryo-EM) to resolve structures essential for immature HIV-1 virus assembly. The system involved tomography acquisition schemes implemented by WHJH, with technical support from the EMBL IT services unit and the EMBL proteomics core facility. Samples were prepared by FKMS and MO, and representative tomograms and final EM structures have been deposited in the EMDB, while the refined HIV-1 CA-SP1 model is in the PDB.'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import bs4\n",
    "from langchain import hub\n",
    "from langchain_chroma import Chroma\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "from langchain_core.output_parsers import StrOutputParser\n",
    "from langchain_core.runnables import RunnablePassthrough\n",
    "from langchain_text_splitters import RecursiveCharacterTextSplitter\n",
    "from langchain_huggingface import HuggingFaceEmbeddings\n",
    "os.environ[\"TOKENIZERS_PARALLELISM\"]=\"false\"\n",
    "# Load, chunk and index the contents of the blog.\n",
    "pdf = '/home/v-haoguan/yuxuan/scihub.py/scihub/EMPIAR-10164.pdf'\n",
    "loader = PyPDFLoader(\n",
    "file_path=pdf,\n",
    ")\n",
    "docs = loader.load()\n",
    "\n",
    "text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)\n",
    "splits = text_splitter.split_documents(docs)\n",
    "from langchain_openai import OpenAIEmbeddings\n",
    "# embedding = OpenAIEmbeddings(base_url=\"https://api.wlai.vip/v1\", api_key=os.environ[\"Ir_API_KEY\"])\n",
    "vectorstore = Chroma.from_documents(documents=splits, embedding=HuggingFaceEmbeddings(model_name=\"infgrad/stella_en_1.5B_v5\"))\n",
    "# vectorstore = Chroma.from_documents(documents=splits, embedding=embedding)\n",
    "\n",
    "# Retrieve and generate using the relevant snippets of the blog.\n",
    "retriever = vectorstore.as_retriever()\n",
    "from langchain.prompts import PromptTemplate\n",
    "\n",
    "# Modify the prompt template\n",
    "custom_prompt_template = PromptTemplate(\n",
    "    input_variables=[\"context\", \"question\"],\n",
    "    template=\"\"\"\n",
    "You are an assistant for question-answering tasks. Use the following pieces of retrieved context to answer the question. Focus on the sections labeled 'Methods' and 'Results' for key details. Provide a detailed description focusing on system setup, resources used, specific characteristics observed, preparations made, and any additional relevant findings.\n",
    "\n",
    "Context:\n",
    "{context}\n",
    "\n",
    "Question:\n",
    "{question}\n",
    "\n",
    "Answer:\n",
    "\"\"\"\n",
    ")\n",
    "\n",
    "# Use the custom prompt in your RAG chain\n",
    "# prompt = custom_prompt_template\n",
    "prompt = hub.pull(\"rlm/rag-prompt\")\n",
    "\n",
    "def format_docs(docs):\n",
    "    return \"\\n\\n\".join(doc.page_content for doc in docs)\n",
    "\n",
    "rag_chain = (\n",
    "    {\"context\": retriever | format_docs, \"question\": RunnablePassthrough()}\n",
    "    | prompt\n",
    "    | llm\n",
    "    | StrOutputParser()\n",
    ")\n",
    "\n",
    "# Invoke with the specific question\n",
    "# docs \n",
    "# retriever.invoke(\"summarize the article for the obtained tomography's caption. It must contain the system, the resources , the characteristics, and the preparations. And more information can throughly describe the tomography. It's the direct information more important, not the summarized and induced supposition and conclusion\")\n",
    "rag_chain.invoke(\"summarize the article for the obtained tomography's caption. It must contain the system, the resources , the characteristics, and the preparations. And more information can throughly describe the tomography. It's the direct information more important, not the summarized and induced supposition and conclusion\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Response: 1. **System Description**:\n",
      "   - The sample comprises β-amyloid and tau proteins from a postmortem Alzheimer's disease (AD) brain. Specifically, it involves extracellular β-amyloid plaques and intracellular and extracellular tau deposits within the mid-temporal gyrus of the brain.\n",
      "\n",
      "2. **Scale**:\n",
      "   - The imaging details includes features on the scale of nanometers. β-amyloid fibrils and tau filaments are measured to be between 3–12 nm in diameter.\n",
      "\n",
      "3. **Focus Point Description**:\n",
      "   - Focus is on β-amyloid plaques and tau inclusions. β-amyloid plaques are composed of fibrils, branched structures, protofilament-like rods, extracellular vesicles, cuboidal particles, and other macromolecular constituents. Tau inclusions are composed of unbranched tau filaments forming clusters.\n",
      "\n",
      "4. **Imaging Environment and Preparation Method**:\n",
      "   - Samples were obtained from rapid autopsy postmortem brain tissue, including a 6-hour postmortem interval. After thawing, acute brain slices were labeled with the fluorescent dye methoxy-X04 (MX04) and vitrified using high-pressure freezing techniques at -188°C. The entire imaging was conducted under cryogenic conditions.\n",
      "\n",
      "5. **Data Acquisition Conditions**:\n",
      "   - Data were acquired using various advanced cryo-electron microscopy techniques including cryo-fluorescence microscopy (cryo-FM) for targeting, cryo-focused ion beam (FIB)-scanning electron microscopy (SEM) for lift-out and thinning, and cryo-electron tomography (cryoET) for three-dimensional imaging. The setup involved a Thermo Fisher Titan Krios 300 keV microscope with a Falcon4i detector and Selectris energy filter. Tomographic tilt series were collected from -60° to +60° with 2° increments.\n",
      "\n",
      "6. **Resolution and Minimum Pixel Size**:\n",
      "   - The resolution of the subtomogram averaging varied between 8.7 angstroms to 31.8 angstroms, depending on the tau filament cluster. The minimum pixel size recorded was 2.38 Å.\n",
      "\n",
      "7. **Labeling and Staining (if applicable)**:\n",
      "   - The amyloid deposits within the tissue were labeled using the fluorescent dye methoxy-X04 (MX04), which binds generically to amyloid. No additional contrasting stains were mentioned.\n",
      "\n",
      "8. **Image Processing and Reconstruction**:\n",
      "   - For image processing, dose-fractionated video frames were aligned, and tilt series were fine-aligned using patch tracking. Initial image processing and subtomogram averaging of tau filaments and Aβ fibrils were performed using PEET (Particle Estimation for Electron Tomography) and further refined using RELION with helical 3D auto-refinement and classification. Helical subtomogram averaging for detailed tau filament structure involved Warp, IMOD, and additional software like EM placement in ChimeraX and Phenix for model fitting and refinement.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import getpass\n",
    "import requests\n",
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "\n",
    "# Prompt for API key securely\n",
    "\n",
    "# Load environment variables\n",
    "azure_endpoint = \"https://gpt4edu-ncus.openai.azure.com/\"\n",
    "azure_deployment = \"GPT4o\"\n",
    "api_version = \"2023-12-01-preview\"\n",
    "api_key = os.environ[\"AZURE_OPENAI_API_KEY\"]\n",
    "\n",
    "# Construct the full URL\n",
    "url = f\"{azure_endpoint}openai/deployments/{azure_deployment}/chat/completions?api-version={api_version}\"\n",
    "\n",
    "# Headers\n",
    "headers = {\n",
    "    \"Content-Type\": \"application/json\",\n",
    "    \"api-key\": f\"{api_key}\"\n",
    "}\n",
    "def format_docs(docs):\n",
    "    return \"\\n\\n\".join(doc.page_content for doc in docs)\n",
    "pdf = '/home/v-haoguan/yuxuan/scihub.py/scihub/EMPIAR-12082.pdf'\n",
    "loader = PyPDFLoader(\n",
    "file_path=pdf,\n",
    ")\n",
    "docs = loader.load()\n",
    "formatted_docs = format_docs(docs)\n",
    "question = f\"\"\"\n",
    "Please extract key information describing Cryo-ET image data from the following article. The information should include a system description, scale, focus point description, imaging environment and preparation method, data acquisition conditions, resolution and minimum pixel size, labeling and staining (if applicable), and image processing and reconstruction. Provide the extracted information in the following format:\n",
    "\n",
    "1. System Description:\n",
    "   - The type of structure or sample shown in the image, such as protein, cell type, and specific site, as well as the chemical or biological environment of the observed system.\n",
    "\n",
    "2. Scale:\n",
    "   - The size of the system and the scale of the information being observed.\n",
    "\n",
    "3. Focus Point Description:\n",
    "   - If it’s a protein, what protein and its site; if it’s a cell, describe the type and specific site.\n",
    "\n",
    "4. Imaging Environment and Preparation Method:\n",
    "   - The environment in which the sample was imaged, such as temperature, natural state, and sample handling.\n",
    "\n",
    "5. Data Acquisition Conditions:\n",
    "   - Type of microscope, accelerating voltage, sample tilt angle range, and other data acquisition parameters.\n",
    "\n",
    "6. Resolution and Minimum Pixel Size:\n",
    "   - The resolution and minimum pixel size of the image.\n",
    "\n",
    "7. Labeling and Staining (if applicable):\n",
    "   - Whether any labeling or staining techniques were used and their types.\n",
    "\n",
    "8. Image Processing and Reconstruction:\n",
    "   - Methods of image processing and 3D reconstruction, such as algorithms or software tools used.\n",
    "\n",
    "Below is the article content:\n",
    "{formatted_docs}\n",
    "\n",
    "Please extract the information and provide it in the specified format.\n",
    "\"\"\"\n",
    "# Request payload\n",
    "payload = {\n",
    "    \"messages\": [\n",
    "        {\"role\": \"system\", \"content\": \"You are a helpful assistant.\"},\n",
    "        {\"role\": \"user\", \"content\": f\"\"\"\n",
    "{question}\n",
    "\n",
    "Answer:\"\"\"\n",
    "         }\n",
    "    ],\n",
    "    # \"max_tokens\": \n",
    "}\n",
    "\n",
    "# Make the request\n",
    "response = requests.post(url, headers=headers, json=payload)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    response_data = response.json()\n",
    "    print(\"Response:\", response_data['choices'][0]['message']['content'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ[\"AZURE_OPENAI_API_KEY\"] = getpass.getpass(prompt=\"Enter your Azure OpenAI API Key: \")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\" \\n  \\n  REPORTS  \\n \\nCite as: Schur  et al., Science  \\n10.1126/science.aaf9620 (2016).  \\n The major structural protein of HIV -1, Gag, oligomerizes at \\nthe plasma membrane of infe cted cells, leading to me m-\\nbrane bending and release of immature virus particles. O r-\\ndered cleavage of the Gag polyprotein at five sites by the viral protease (PR) then causes a dramatic structural rea r-\\nrangement of the virus to the mature, infectious virion (fig. S1, A to C). Gag –Gag interactions in the immature virus are \\nmediated by the CA domains and the adjacent SP1 region. After maturation, CA forms a conical core encapsulating the viral RNA -nucleoprotein complex ( 1). The final proteolytic \\ncleavage in Gag  occurs between CA and SP1. Even small \\nremnants of uncleaved CA -SP1 have a dominant -negative \\neffect on infectivity ( 2, 3). Cleavage at this site is inhibited \\nby MIs ( 4) (fig. S1B), and polymorphisms in this region \\ncaused  resistance against the first -in-class MI Bevirimat \\n(BVM) (5 ). Other MIs inhibiting CA -SP1 cleavage are cu r-\\nrently undergoing clinical trials, but the precise mechanism of their inhibitory action and of resistance is unclear.  \\nUnprocessed Gag is assembled in to irregular curved la t-\\ntices whose structures cannot be determined using conve n-\\ntional structural biology techniques. The best structural model for CA within the immature virus has been derived by fitting available NMR and crystal structures of the folded N-terminal and C -terminal domains (CA -NTD and CA -CTD, \\nrespectively) into a structure of the immature CA lattice de-rived by cryo- electron tomography and subtomogram ave r-\\naging ( 6). Subtomogram averaging can resolve protein structures within complex environmen ts ranging from cells \\nto enveloped viruses ( 7), but has been limited to resolutions \\nof approximately 8 Å. The positions of α-helices, but not of \\namino acids, are visible at this resolution. Furthermore, 8 Å resolution is not sufficient to generate ab initio structural models for unknown structures such as the Gag regions downstream of the crystallized CA -CTD, which ends at res i-\\ndue 220 of the 231 amino acid CA domain (fig. S1D). This downstream region consists of amino acids thought to as-semble a flexible hinge, followed by a sequence predicted to form a six- helix bundle that spans the C -terminal residues of \\nCA, the cleavage site and the N -terminal residues of SP1 ( 8). \\nThis region contains the critical CA -SP1 cleavage site and \\nincludes the majority of amino a cids essential for virus a s-\\nsembly ( 9–11). Obtaining a high -resolution structure of CA -\\nSP1 in the immature arrangement is vital for a mechanistic understanding of HIV -1 assembly, maturation, and its inh i-\\nbition.  \\nThe HIV -1 Gag construct ΔMACANCSP2 ( 11) assemb les \\ninto immature virus -like particles (VLPs) in vitro (Fig. 1A \\nand fig. S1C). We assembled Δ MACANCSP2 particles in the \\nabsence or presence of 100 μ g/ml BVM. We also purified \\nintact, immature HIV -1 virus particles carrying an inactiva t-\\ning PR mutation, D25A . For all three samples we acquired \\ncryo-electron tomography tilt series implementing an opt i-\\nmized data collection scheme ( 12, 13) (table S1). VLPs with \\nor without BVM were indistinguishable, showing densities An atomic model of HIV -1 capsid -SP1 reveals structures \\nregulating assembly and maturation  \\nFlorian K. M. Schur,1,2 Martin Obr,2,3 Wim J. H. Hagen,1 William Wan,1 Arjen J. Jakobi,1,4 Joanna M. \\nKirkpatrick,5* Carsten Sachse,1 Hans- Georg Kräusslich,2,3 John A. G. Briggs1,2† \\n1Structural and Computational Biology Unit, European Molecular Biology Laboratory, Meyerhofstraße 1, 69117 Heidelberg, Germany . 2Molecular Medicine Partnership Unit,  \\nEMBL/Universitätsklinikum Heidelberg, Heidelberg, Germany. 3Department of Infectious Diseases, Virology, Universitätsklinikum Heidelberg, Im Neuenheimer Feld 324, \\n69120 Heidelberg, Germany. 4Hamburg Unit c/o DESY, European Molecular Biology Laboratory, No tkestraße 85, 22607 Hamburg, Germany. 5Proteomics Core Facility, \\nEuropean Molecular Biology Laboratory, Meyerhofstraße 1, 69117 Heidelberg, Germany.  \\n*Present address: Leibniz Institute on Aging –Fritz Lipmann Institute (FPI), Beutenbergstraße 11, 07745 Jena , Germany.  \\n†Corresponding author. Email: john.briggs@embl.de  \\nImmature HIV- 1 assembles at, and buds from, the plasma membrane before proteolytic cleavage of the \\nviral Gag polyprotein induces structural maturation. Maturation is blocked by protease and matur ation \\ninhibitors (MIs), abolishing infectivity. The CA (capsid) and SP1 (spacer peptide 1) region of Gag is the key regulator of assembly and maturation, and the target of MIs. Here we applied optimized cryo -electron \\ntomography and subtomogram averaging to  resolve this region within assembled immature HIV -1 particles \\nat 3.9 Å resolution and built an atomic model. The structure reveals a network of intra - and inter -\\nmolecular interactions mediating immature HIV- 1 assembly. The proteolytic cleavage site betwee n CA and \\nSP1 is inaccessible to protease. We suggest that MIs prevent CA -SP1 cleavage by stabilizing the structure, \\nand MI resistance develops by destabilizing CA -SP1.  \\nFirst release: 14 July  2016   www.sciencemag.org   ( Page numbers not final at time of first  release ) 1   \\n \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \\n\\nfor the Gag lattice similar to the immature virus particles \\n(Fig. 1A). Subtomogram averaging was performed ind e-\\npendently for each dataset using an optimized workflow implementing features including frame -based motion cor-\\nrection ( 14) and exposure filtering ( 15). The resolutions of \\nthe CA -SP1 layer in the final averages of untreated or BVM -\\ntreated VLPs, and immature virus were 4.5, 3.9 and 4.2 Å, respectively (Fig. 1B and fig. S2, A and B).  \\nWe compared the three structures and found no clear \\ndifferences in the protein densities of CA or SP1 – they vary \\nonly in the presence or absence of densities at the center of \\nthe hexamer near SP1 (fig. S2, C to I, and fig. S3C). Given the high degree of structural similarity between the maps we used the 3.9 Å structure to build and refine a complete atomic model for t he CA -SP1 region from Gag residues 148 \\nto 371 (Figs. 1C and 2, figs. S3 and S4, table S2, and movies S1 and S2). Residues 356 -371, covering the C terminus of CA, \\nand the first eight residues of SP1, assemble into a six -helix \\nbundle.  \\nInteractions within the  CA-NTD layer are described in \\nfig. S4A. The CA -NTD and CA -CTD contact one another in \\nthe region of E160 -E161 in the CA- NTD and Q308 in the CA -\\nCTD. The relative positions of the two domains may also be restricted by the extended, rigid linker between them,  which \\nappears to be stabilized by an interaction with R305 in helix 8. Y277, the last residue of helix 7 in a solution structure ( 16) \\nand in the mature -like CA hexamer ( 17), is rotated out of the \\nhelix, and packs against the linker at P279 (Fig. 2B). This network of interactions may allow the CA -CTD to be stru c-\\nturally modulated by cleavage upstream of CA -NTD and vice \\nversa.  \\nHighly conserved residues of the Major Homology R e-\\ngion (MHR) in CA -CTD abut the extended chain between \\nthe 3\\n10-helix and helix 8 (Fig.  2C). Charged residues in the \\n310-helix and extended chain make interactions with the \\nneighboring CA -CTD molecule within the hexamer (Fig. 2, B \\nand C). Hexamers are linked by a CA -CTD dimer interface \\n(fig. S4B) formed by residues W316 and M317 ( 6, 18, 19). \\nDownstream of helix 11, beyond the residues that are re-\\nsolved in crystal structures ( 19, 20), the “hinge region” VGG \\n(residues 353 -355) unexpectedly adopts a rigid structure \\nwithin the lattice. P356 then marks the start of a helix (r e-\\nferred to here as the  CA-SP1 helix) that extends down to re s-\\nidue 371 in SP1 (Fig. 2, A and D) before abruptly ending, indicating that residues C -terminal of 371 are disordered \\n(movie S1). The CA -SP1 helix protrudes up into the CA -CTD \\nlayer, where the top of the helix and the VGG “hinge” from one Gag molecule pack tightly against the CA- CTD from the \\nneighboring Gag molecule. The CA -CTD, VGG hinge and CA -\\nSP1 helix thus form a single, integrated “assembly unit” that defines the structure of the hexamer (Fig. 2, D and G).  \\nThis asse mbly unit appears to be stabilized by a three -way interaction between H358 in the CA- SP1 helix, D329 in \\nthe base of helix 10, and P356 from the CA -SP1 helix in the \\nneighboring CA molecule (Fig. 2E). K290 in the loop u p-\\nstream of helix 8, and K359 in the CA -SP1 helix protrude \\nfrom above and below these residues toward the center of the six- helix bundle where they coordinate a density, pr e-\\nsumably a negatively charged ion cluster (Fig. 2F and movie S2). This arrangement is reminiscent of six arginine res i-\\ndues t hat protrude into the center of the NTD hexamer in \\nmature HIV capsid ( 17, 21). 12 essential amino residues in \\nCA-CTD have been identified where mutation to alanine \\nabolishes virus assembly ( 22, 23). These are: W316 and M317 \\nin the hydrophobic dimeric interface; V353, G354 and G355 in the VGG hinge; K290, D329, P356, H358 and K359, which together form the intricate network of interactions that d e-\\nfine the assembly unit; and A360 and L363 which appear to make hydrophobic interactions within the six -helix bundl e. \\nThere is therefore a remarkably close correlation between the sensitivity of a residue to mutation and a role for the residue in mediating interactions within the CA -SP1 asse m-\\nbly unit, confirming the importance of these interactions in virus assembly.  \\nDuring maturation, the final proteolytic cleavage occurs \\nbetween L363 and A364. In our structure, this site is in the middle of the CA -SP1 helix bundle where it is inaccessible to \\nPR, which acts on extended protein chains (24 ) (Fig. 2D and \\nmovie S2). Disass embly of the immature lattice and full \\ncleavage between CA and SP1 only take place once cleavage has occurred both between MA and CA, and between SP1 and NC ( 10, 25). Together, these observations support a \\nmodel where the final step in maturation is regulated by limiting the access of PR to its substrate: cleavage upstream of CA and downstream of SP1 together destabilize the im-mature lattice and the CA -SP1 helix, thereby allowing PR to \\ncleave between CA and SP1.  \\nMIs block HIV infection by preventing cleavage at the \\nCA-SP1 site ( 4, 13, 26) (fig. S5), and may also stabilize the \\nimmature lattice ( 27, 28). We mapped mutations, deletions \\nand polymorphisms that confer resistance to BVM ( 5, 29, \\n30) and to another MI, PF -46396 ( 31, 32) onto our structural \\nmodel (Fig. 3). This revealed that resistance mutations do not map out potential drug -binding pockets. Instead, they \\nare located at protein -protein interfaces within the CA -SP1 \\nlattice. Together with our observation that the CA -SP1 \\ncleavage site is inaccessible in the immature virus, this im-\\nplies that the mode of action of MIs is not steric inhibition of proteolysis but instead stabilization of the immature Gag lattice. Our data suggest BVM stabilizes the lattice by bin d-\\ning to a site in the center of the six- helix bundle (fig. S2). \\nThe reduced cleavage at the CA -SP1 boundary is a dow n-\\nstream effect of stabilization because cleavage of this site requires unfolding to an extended chain. HIV -1 appears to \\nFirst release: 14 July  2016  www.sciencemag.org   ( Page numbers not final at time of first  release ) 2 \\n \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \\n\\ndevelop MI resistance by destabilizi ng its immature form, \\nthus directly counteracting the stabilizing effects of MIs.  \\nREFERENCES AND NOTES  \\n1. W. I. Sundquist, H. G. Kräusslich, HIV -1 assembly, budding, and maturation. Cold \\nSpring Harb. Perspect. Med.  2, a006924 (2012).  Medline  \\ndoi:10.1101/cshperspect.a015420  \\n2. B. Müller, M. Anders, H. Akiyama, S. Welsch, B. Glass, K. Nikovics, F. Clavel, H. -M. \\nTervo, O. T. Keppler, H. -G. Kräusslich, HIV- 1 Gag processing intermediates trans -\\ndominantly interfere with HIV- 1 infectivity. J. Biol. Chem.  284 , 29692 –29703 \\n(2009).  Medline  doi:10.1074/jbc.M109.027144  \\n3. M. A. Checkley, B. G. Luttge, F. Soheilian, K. Nagashima, E. O. Freed, The capsid -\\nspacer peptide 1 Gag processing intermediate is a dominant -negative inhibitor of \\nHIV-1 maturation. Virology  400 , 137 –144 (2010).  Medline  \\ndoi:10.1016/j.virol.2010.01.028  \\n4. C. S. Adamson, K. Salzwedel, E. O. Freed, Virus maturation as a new HIV -1 \\ntherapeutic target. Expert Opin. Ther. Targets 13, 895 –908 (2009).  Medline  \\ndoi:10.1517/14728220903039714  \\n5. C. S. Adamson, M. Sakalian, K. Salzwedel, E. O. Freed, Polymorphisms in Gag  \\nspacer peptide 1 confer varying levels of resistance to the HIV - 1maturation \\ninhibitor bevirimat. Retrovirology  7, 36 (2010). doi:10.1186/1742 -4690 -7-36 \\n6. F. K. Schur, W. J. Hagen, M. Rumlová, T. Ruml, B. Müller, H. G. Kräusslich, J. A. \\nBriggs, Structure of the immature HIV -1 capsid in intact virus particles at 8.8 Å \\nresolution. Nature  517, 505–508 (2015).  Medline  doi:10.1038/nature13838  \\n7. J. A. Briggs, Structural biology in situ —the potential of subtomogram averaging. \\nCurr. Opin. Struct. Biol.  23, 261 –267 (2013).  Medline  \\ndoi:10.1016/j.sbi.2013.02.003  \\n8. M. A. Accola, S. Höglund, H. G. Göttlinger, A putative α-helical structure which \\noverlaps the capsid -p2 boundary in the human immunodeficiency virus type 1 \\nGag precursor is crucial for viral particle assembly. J. Virol.  72, 2072 –2078 \\n(1998).  Medline  \\n9. S. A. Datta, L. G. Temeselew, R. M. Crist, F. Soheilian, A. Kamata, J. Mirro, D. \\nHarvin, K. Nagashima, R. E. Cachau, A. Rein, On the role of the SP1 domain in \\nHIV-1 particle assembly: A molecular switch? J. Virol.  85, 4111– 4121 (2011).  \\nMedline  doi:10.1128/JVI.00006 -11 \\n10. A. d e Marco, B. Müller, B. Glass, J. D. Riches, H. -G. Kräusslich, J. A. G. Briggs, \\nStructural analysis of HIV -1 maturation using cryo -electron tomography. PLOS \\nPathog.  6, e1001215 (2010). doi:10.1371/journal.ppat.1001215  \\n11. I. Gross, H. Hohenberg, T. Wilk, K. Wiegers, M. Grättinger, B. Müller, S. Fuller, H. G. \\nKräusslich, A conformational switch controlling HIV -1 morphogenesis. EMBO J.  \\n19, 103 –113 (2000).  Medline  doi:10.1093/emboj/19.1.103  \\n12. W. J. Hagen, W. Wan, J. A. Briggs, Implementation of a cryo -electron tomography \\ntilt-scheme optimized for high resolution subtomogram averaging. J. Struct. Biol.  \\n10.1016/j.jsb.2016.06.007 (2016).  Medl ine doi:10.1016/j.jsb.2016.06.007  \\n13. Materials and methods are available as supplementary materials on Science  \\nOnline.  \\n14. X. Li, P. Mooney, S. Zheng, C. R. Booth, M. B. Braunfeld, S. Gubbens, D. A. Agard, \\nY. Cheng, Electron counting and beam -induced motion correction enable near -\\natomic- resolution single- particle cryo -EM. Nat. Methods  10, 584 –590 (2013).  \\nMedline  doi:10.1038/nmeth.2472  \\n15. T. Grant, N. Grigorieff, Measuring the optimal exposure for single particle cryo -\\nEM using a 2.6 Å reconstruction of rotavirus VP6. eLife  4, e06980 (2015).  \\nMedline  doi:10.7554/eLife.06980  \\n16. C. Tang, Y. Ndassa, M. F. Summers, Structure of the N -terminal 283 -residue \\nfragment of the immature HIV -1 Gag polyprotein. Nat. Struct. Biol.  9, 537 –543 \\n(2002).  Medline  \\n17. A. T. Gres, K. A. Kirby, V. N. KewalRamani, J. J. Tanner, O. Pornillos, S. G. \\nSarafianos, X -ray crystal structures of native HIV- 1 capsid protein reveal \\nconformational variability. Science  349 , 99 –103 (2015).  Medline  \\ndoi:10.1126/science.aaa5936  \\n18. J. A. Briggs, J. D. Riches, B. Glass, V. Bartonova, G. Zanetti, H. G. Kräusslich, \\nStru cture and assembly of immature HIV. Proc. Natl. Acad. Sci. U.S.A. 106 , \\n11090 –11095 (2009).  Medline  doi:10.1073/pnas.0903535106  \\n19. V. Bartonova, S. Igonet, J. Sticht, B. Glass, A. Habermann, M. C. Vaney, P. Sehr, J. \\nLewis, F. A. Rey, H. G. Kraüsslich, Residues in the HIV -1 capsid assembly inhibitor binding site are essential for maintaining the assembly- competent quaternary \\nstructure of the capsid protein. J. Biol. Chem.  283 , 32024 –32033 (2008).  \\nMedline  doi:10.1074/jbc.M804230200  \\n20. T. R. Gamble, S. Yoo, F. F. Vajdos, U. K. von Schwedler, D. K. Worthylake, H. \\nWang, J. P. McCutcheon, W. I. Sundquist, C. P. Hill, Structure of the carboxyl-terminal dimerization domain of the HIV -1 capsid protein. Science \\n278 , 849– 853 \\n(1997).  Medline  doi:10.1126/science.278.5339.849  \\n21. O. Pornillos, B. K. Ganser -Pornillos, B. N. Kelly, Y. Hua, F. G. Whitby, C. D. Stout, \\nW. I. Sundquist, C. P. Hill, M. Yeager, X -ray structures of the hexameric building \\nblock of the HIV capsid. Cell  137, 1282 –1292 (2009).  Medline  \\ndoi:10.1016/j.cell.2009.04.063  \\n22. U. K. von Schwedler, K. M. Stray, J. E. Garrus, W. I. Sundquist, Functional \\nsurfaces of the human immunodeficiency virus type 1 capsid protein. J. Virol.  77, \\n5439 –5450 (2003).  Medline  doi:10.1128/JVI.77.9.5439 -5450.2003  \\n23. D. Melamed, M. Mark -Danieli, M. Kenan- Eichler, O. Kraus, A. Castiel, N. Laham, T. \\nPupko, F. Glaser, N. Ben -Tal, E. Bacharach, The conserved carboxy terminus of \\nthe capsid domain of human immunodeficiency virus type 1 Gag protein is important for virion assembly and release. J. Virol.  \\n78, 9675 –9688 (2004).  \\nMedline  doi:10.1128/JVI.78.18.9675 -9688.2004  \\n24. M. Prabu -Jeyabalan, E. Nalivaika, C. A. Schiffer, Substrate shape determines \\nspecificity of recognition for HIV -1 protease: Analysis of crystal structures of six \\nsubstrate complexes. Structure  10, 369 –381 (2002).  Medline  \\ndoi:10.1016/S0969 -2126(02)00720 -7 \\n25. K. Wiegers, G. Rutter, H. Kottler, U. Tessmer, H. Hohenberg, H. -G. Kräusslich, \\nSequential steps in human immunodeficiency virus particle maturation revealed by alterations of individual Gag polyprotein cleavage sites. J. Virol.  \\n72, 2846 –\\n2854 (1998).  Medline  \\n26. M. Sakalian, C. P. McMurtrey, F. J. Deeg, C. W. Maloy, F. Li, C. T. Wild, K. \\nSalzwedel, 3 -O-(3′,3′ -dimethysuccinyl) betulinic acid inhibits maturation of the \\nhuman immunodeficiency virus type 1 Gag precursor assembled in vitro. J. Virol.  \\n80, 5716 –5722 (2006). doi:10.1128/JVI.02743 -05 \\n27. P. W. Keller, R. K. Huang, M. R. England, K. Waki, N. Cheng, J. B. Heymann, R. C. \\nCraven, E. O. Freed, A. C. Steven, A two -pronged structural analysis of retroviral \\nmaturation indicates that core formation proceeds by a disassembly -reassembly \\npathway rather than a displacive transition. J. Virol.  87, 13655 –13664 (2013).  \\nMedline  doi:10.1128/JVI.01408 -13 \\n28. P. W. Keller, C. S. Adamson, J. B. Heymann, E. O. Free d, A. C. Steven, HIV -1 \\nmaturation inhibitor bevirimat stabilizes the immature Gag lattice. J. Virol.  85, \\n1420–1428 (2011).  Medline  doi:10.1128/JVI.01926 -10 \\n29. C. S. Adamson, S. D. Ablan, I. Boeras, R. Goila -Gaur, F. Soheilian, K. Nagashima, \\nF. Li, K. Salzwedel, M. Sakalian, C. T. Wild, E. O. Freed, In vitro resistance to the human immunodeficiency virus type 1 maturation inhibitor PA -457 (bevirimat). \\nJ. Virol.  \\n80, 10957 –10971 (2006).  Medline doi:10.1128/JVI.01369 -06 \\n30. N. A. Margot, C. S. Gibbs, M. D. Miller, Phenotypic susceptibility to bevirimat in \\nisolates from HIV -1-infected patients without prior exposure to bevirimat. \\nAntimicrob. Agents Chemother.  54, 2345 –2353 (2010).  Medline  \\ndoi:10.1128/AAC.01784 -09 \\n31. K. Waki, S. R. Durell, F. Soheilian, K. Nagashima, S. L. Butler, E. O. Freed, \\nStructural and functional insights into the HIV -1 maturation inhibitor binding \\npocket. PLOS Pathog.  8, e1002997 (2012).  Medline  \\ndoi:10.1371/journal.ppat.1002997  \\n32. W. S. Blair, J. Cao, J. Fok -Seang, P. Griffin, J. Isaacson, R. L. Jackson, E. Murray, \\nA. K. Patick, Q. Peng, M. Perros, C. Pic kford, H. Wu, S. L. Butler, New small-\\nmolecule inhibitor class targeting human immunodeficiency virus type 1 virion maturation. Antimicrob. Agents Chemother.  \\n53, 5080 –5087 (2009).  Medline  \\ndoi:10.1128/AAC.00759 -09 \\n33. M. Dettenhofer, X.- F. Yu, Highly purified human immunodeficiency virus type 1 \\nreveals a virtual absence of Vif in virions . J. Virol.  73, 1460 –1467 (1999).  Medline  \\n34. D. N. Mastronarde, Automated electron microscope tomography using robust \\nprediction of specim en movements. J. Struct. Biol.  152, 36 –51 (2005).  Medline  \\ndoi:10.1016/j.jsb.20 05.07.007  \\n35. M. Vulović, E. Franken, R. B. G. Ravelli, L. J. van Vliet, B. Rieger, Precise and \\nunbiased estimation of astigmatism and defocus in transmission electron microscopy. Ultramicroscopy  \\n116, 115 –134 (2012). \\ndoi:10.1016/j.ultramic.2012.03.004  \\nFirst release: 14 July  2016  www.sciencemag.org   ( Page numbers not final at time of first  release ) 3 \\n \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \\n\\n36. A. Rohou, N. Grigorieff, CTFFIND4: Fast and accurate defocus estimation from \\nelectron micrographs. J. Struct. Biol.  192 , 216 –221 (2015).  Medline  \\ndoi:10.1016/j.jsb.2015.08.008  \\n37. F. K. Schur, W. J. Hagen, A. de Marco, J. A. Briggs, Determination of protein \\nstructure at  8.5Å resolution using cryo -electron tomography and sub -tomogram \\naveraging. J. Struct. Biol.  184 , 394 –400 (2013).  Medline  \\ndoi:10.1016/j.jsb.2013.10.015  \\n38. J. R. Kremer, D. N. Mastronarde, J. R. McIntosh, Computer visualization of three -\\ndimensional image data using IMOD. J. Struct. Biol.  116, 71–76 (1996).  Medline  \\ndoi:10.1006/jsbi.1996.0013  \\n39. Q. Xiong, M. K. Morphew, C. L. Schwartz, A. H. Hoenger, D. N. Mastronarde, CTF \\ndetermination and correction for low dose tomographic tilt series. J. Struct. Biol.  \\n168 , 378–387 (2009).  Medline  doi:10.1016/j.jsb.2009.08.016  \\n40. S. Pruggnaller, M. Mayr, A. S. Frangakis, A visualization and segmentation \\ntoolbox for electron microscopy. J. Struct. Biol.  164 , 161 –165 (2008).  Medline  \\ndoi:10.1016/j.jsb.2008.05.003  \\n41. F. Förster, O. Medalia, N. Zauberman, W. Baumeister, D. Fass , Retrovirus \\nenvelope protein complex structure in situ studied by cryo -electron tomography. \\nProc. Natl. Acad. Sci. U.S.A.  102 , 4729 –4734 (2005).  Medline  \\ndoi:10.1073/pnas.0409178102  \\n42. S. Nickell, F. Förster, A. Linaroudis, W. D. Net, F. Beck, R. Hegerl, W. Baumeister, \\nJ. M. Plitzko, TOM software toolbox: Acquisition and analysis for  electron \\ntomography. J. Struct. Biol.  149 , 227 –234 (2005).  Medline  \\ndoi:10.101 6/j.jsb.2004.10.006  \\n43. D. Castaño- Díez, M. Kudryashev, M. Arheit, H. Stahlberg, Dynamo: A flexible, \\nuser -friendly development tool for subtomogram averaging of cryo -EM data in \\nhigh- performance computing environments. J. Struct. Biol.  178, 139 –151 (2012).  \\nMedline  doi:10.1016/j.jsb.2011.12.017  \\n44. S. Chen, G. McMullan, A. R. Faruqi, G. N. Murshudov, J. M. Short, S. H. W. \\nScheres, R. Henderson, High -resolution noise substitution to measure overfitting \\nand validate resolution in 3D structure determination by single particle electron \\ncryomicroscopy. Ultramicroscopy  135 , 24 –35 (2013).  Medline  \\ndoi:10.1016/j.ultramic.2013.06.004  \\n45. P. B. Rosenthal, R. Henderson , Optimal determination of particle orientation, \\nabsolute hand, and contrast loss in single -particle electron cryomicroscopy. J. \\nMol. Biol.  333 , 721 –745 (2003).  Medline  doi:10.1016/j.jmb.2003.07.013  \\n46. E. F. Pettersen, T. D. Goddard, C. C. Huang, G. S. Couch, D. M. Greenblatt, E. C. \\nMeng, T. E. Ferrin, UCSF Chimera —a visualization system for exploratory \\nresearch and analysis. J. Comput. Chem.  25, 1605 –1612 (2004).  Medline  \\ndoi:10.1002/jcc.20084  \\n47. P. Emsley, K. Cowtan, Coot: Model -building tools for molecular graphics. Acta \\nCrystallogr. D Biol. Crystallogr.  60, 2126 –2132 (2004).  Medline  \\ndoi:10.1107/S0907444904019158  \\n48. The PyMOL Molecular Graphics System (Schrodinger, 2010).  \\n49. N. Morellet, S. Drui llennec, C. Lenoir, S. Bouaziz, B. P. Roques, Helical structure \\ndetermined by NMR of the HIV- 1 (345 –392)Gag sequence, surrounding p2: \\nImplications for particle assembly and RNA packaging. Protein Sci.  14, 375 –386 \\n(2005).  Medline  doi:10.1110/ps.041087605  \\n50. N. A. Hoffmann, A. J. Jakobi, M. Moreno -Morcillo, S. Glatt, J. Kosinski, W. J. H.  \\nHagen, C. Sachse, C. W. Müller, Molecular structures of unbound and transcribing RNA polymerase III. Nature  \\n528 , 231 –236 (2015).  Medline  \\ndoi:10.1038/nature16143  \\n51. S. A. Fromm, T. A. Bharat, A. J. Jakobi, W. J. Hagen, C. Sachse, Seeing tobacco \\nmosaic virus through direct electron detectors. J. Struct. Biol.  189 , 87 –97 \\n(2015).  Medline  doi:10.1016/j.jsb.2014.12.002  \\n52. M. D. Winn, C. C. Ballard, K. D. Cowtan, E. J.  Dodson, P. Emsley, P. R. Evans, R. M. \\nKeegan, E. B. Krissinel, A. G. W. Leslie, A. McCoy, S. J. McNicholas, G. N. Murshudov, N. S. Pannu, E. A. Potterton, H. R. Powell, R. J. Read, A. Vagin, K. S. Wilson, Overview of the CCP4 suite and current development s. Acta Crystallogr. \\nD Biol. Crystallogr.  \\n67, 235 –242 (2011).  Medline  \\ndoi:10.1107/S0907444910045749  \\n53. P. D. Adams, P. V. Afonine, G. Bunkóczi, V. B. Chen, I. W. Davis, N. Echols, J. J. \\nHeadd, L. -W. Hung, G. J. Kapral, R. W. Grosse -Kunstleve, A. J. McCoy, N. W. \\nMoriarty, R. Oeffner, R. J. Read, D. C. Richardson, J. S. Richardson, T. C. Terwilliger, P. H. Zwart, PHENIX: A comprehensive Python -based system for macromolecular structure solution. Acta Crystallogr. D Biol. Crystallogr.  \\n66, 213 –\\n221 (2010).  Medline  doi:10.1107/S0907444909052925  \\n54. R. W. Grosse -Kunstleve, N. K. Sauter , N. W. Moriarty, P. D. Adams, The \\nComputational Crystallography Toolbox: Crystallographic algorithms in a reusable software framework. J. Appl. Cryst.  \\n35, 126 –136 (2002). \\ndoi:10.1107/S00218898010 17824  \\n55. C. J. Williams, B. J. Hintze, D. C. Richardson, J. S. Richardson, CaBLAM: \\nIdentification and scoring of disguised secondary structure at low resolution. Comput. Crystallogr. Newsl.  \\n4, 33 (2013).  \\n56. C. Colliex, J. M. Cowley, S. L. Dudarev, M. Fink, J. Gjønnes, R. Hilderbrandt, A. \\nHowie, D. F. Lynch, L. M. Peng, G. Ren, A. W. Ross, V. H. Smith, J. C. H. Spence, J. W. Steeds, J. Wang, M. J. Whelan, B. B. Zvyagin, “Electron diffraction,” in International Tables for Crystallography, Volume C: Mathematical, Physical, and \\nChemical Tables , E. Prince, Ed. (International Union of Crystallography, 2006), \\nchap. 4.3, pp. 259 –429. \\n57. V. B. Chen, W. B. Arendall 3rd, J. J. Headd, D. A. Keedy, R. M. Immormino, G. J. \\nKapral, L. W. Murray, J. S. Richardson,  D. C. Richardson, MolProbity: All- atom \\nstructure validation for macromolecular crystallography. Acta Crystallogr. D Biol. Crystallogr.  \\n66, 12–21 (2010).  Medline  doi:10.1107/S0907444909042073  \\n58. T. A. Bharat, L. R. Castillo Menendez, W. J. Hagen, V. Lux, S. Igonet, M. Schorb, F. \\nK. Schur, H. G. Kräusslich, J. A. Briggs, Cryo -electr on microscopy of tubular \\narrays of HIV -1 Gag resolves structures essential for immature virus assembly. \\nProc. Natl. Acad. Sci. U.S.A.  111, 8233 –8238 (2014).  Medline  \\ndoi:10.1073/pnas.1401455111  \\nACKNOWLEDGMENTS  \\nThe authors thank B. Glass for preparation of the immature HIV -1 (D25A) sample, \\nJ.Plitzko and D.Tegunov for providing the K2Align software, and S.Mattei, N. Hoffman, F.Thommen, A.Sonnen and S.Dodonova for technical assistance and/or discussion. This study was supported by Deutsche Forschungsgemeinschaft grants BR 3635/2 -1 to JAGB and KR 906/7 -1 to H- GK. \\nThe Briggs laboratory acknowledges financial support from the European Molecular Biology Laboratory (EMBL) and from the Chica und Heinz Schaller Stiftung. WW was supported by an EMBO Long Term Fellowship (ALTF 748 -\\n2014). AJJ acknowledges support by the EMBL Interdisciplinary Postdoc  \\nProgram (EIPOD) under Marie Curie COFUND actions (PCOFUND- GA-2008 -\\n229597) and the Joachim Herz Stiftung. This study was technically supported by the EMBL IT services unit and the EMBL proteomics core facility. FKMS, MO, H-GK and JAGB designed the experiments. FKMS and MO prepared samples. JMK helped designing the MS experiments. WHJH implemented tomography acquisition schemes. FKMS and WHJH acquired the data. FKMS and WW performed image processing. FKMS, AJJ and CS refined the model. FKMS, MO and JAGB analyzed the data. FKMS and JAGB wrote the manuscript with support from all authors. Representative tomograms, and the final electron microscopy (EM) structures have been deposited in the EMDB with accesssion numbers EMD -4015, EMD -4016, EMD- 4017, EMD -4018, EMD -4019 and EMD -4020. The \\nrefined HIV -1 CA -SP1 model has been deposited in the PDB with accession \\nnumber 5L93. \\nSUPPLEMENTARY MATERIALS  \\nwww.sciencemag.org/cgi/content/full/science.aaf9620/DC1  \\nMaterials and Methods  \\nFigs. S1 to S5  \\nTables S1 to S2  \\nReferences ( 33– 58) \\nMovies S1 to S2  \\n 26 April 2016; accepted 29 June 2016  \\nPublished online 14 July 2016  \\n10.1126/science.aaf9620  \\n \\nFirst release: 14 July  2016  www.sciencemag.org   ( Page numbers not final at time of first  release ) 4 \\n \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \\n\\n \\n  \\nFig. 1. Structure of the immature HIV -1 CA -SP1 lattice at 3.9 Å.  \\n(A) Computational slices through tomograms containing either \\nimmature HIV -1 or untreated or BVM treated ΔMACANCSP2 VLPs. \\nArrowhead marks the membrane bilayer in the left hand panel. Scale \\nbar 50 nm. ( B) Electron densities of CA -SP1 viewed perpendicular to \\nthe lattice, generated by subtomogram averaging from the samples \\nshown in (A). One CA -SP1 monomer is highlighted in color with the \\nCA- NTD and CA -CTD/SP1 in cyan and orange, respectively. The \\nrespective resolutions of the determined structures are annotated. (\\nC) The refined atomic model shown in the same view as in (B) and \\nrotated by 90°, viewed from outside of the virus. The 6- fold \\nsymmetry axis is annotated with a hexagon.  \\n \\nFirst release: 14 July  2016  www.sciencemag.org   ( Page numbers not final at time of first  release ) 5 \\n \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \\n\\n  \\n  \\nFig. 2. Structural features in the immature CA -SP1 lattice.  (A) A single CA -SP1 monomer as in Fig. 1C. α -\\nhelices and the Cyclophilin A binding loop are annotated ( α-helices from neighboring CA monom ers are \\nannotated in brackets). Colored rectangles indicate regions enlarged in (B) to (D). ( B) The CA -NTD/CA -CTD \\nlinker is in an extended conformation with Y277 binding to the linker and S278 approaching R305. ( C) The \\nhighly conserved residues in the MHR (Q287, E291, R299) stabilize the linker connecting the 3 10-helix and \\nhelix 8. Residues in this linker can interact with an adjacent CA monomer around the hexameric ring, e.g., \\nR286 with E344,  and D284 with R294; point mutations of these residues do not abolish assembly ( 22), \\nsuggesting some redundancy in these interactions. ( D) The CA -CTD, VGG hinge, and the top of the CA -SP1 \\nhelix form an integrated structural assembly unit. The CA -SP1 cleava ge site is marked by a blue asterisk. \\nDashed rectangles indicate approximate position of enlarged panels (E) and (F). Residues colored as in (E) and (F). (\\nE) Residues D329, P356 and H358 (in purple) form a three way linkage between two neighboring CA -\\nSP1 h elices and the base of the CA -CTD. ( F) K290 and K359 (in green) protrude from above and below the \\nregion shown in (E), to the center of the hexamer where they coordinate a strong density. ( G) Horizontal (left) \\nand vertical (right) slabs through the structure illustrate that the MHR (yellow), other residues in the CA -CTD \\nbase (red), the VGG hinge (blue) and the top of the CA -SP1 helix (pink) come together to form the hexameric \\nassembly unit. The vertical slab shows one half of the hexamer represented in a surface view.  \\n \\nFirst release: 14 July  2016  www.sciencemag.org   ( Page numbers not final at time of first  release ) 6 \\n \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \\n\\n \\n \\n \\nFig. 3. Maturation inhibitor resistance mutations \\ndestabilize the immature lattice.  Resistance mutations or \\nnaturally occurring polymorphisms that render HIV -1 \\nresistant to BVM (red) or PF -46396 (green) have been \\nmapped onto the atomic model. Mutations that confer \\nresistance for both compounds are colored yellow. The CA -\\nSP1 cleavage site is marked by a blue asterisk. For clarity \\nonly three helices of the six -helix bundle are displayed.  \\nFirst release: 14 July  2016  www.sciencemag.org   ( Page numbers not final at time of first  release ) 7 \\n \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \\n\\npublished online July 14, 2016Kräusslich and John A. G. Briggs (July 14, 2016) J. Jakobi, Joanna M. Kirkpatrick, Carsten Sachse, Hans-Georg Florian K. M. Schur, Martin Obr, Wim J. H. Hagen, William Wan, Arjenassembly and maturationAn atomic model of HIV-1 capsid-SP1 reveals structures regulating\\n\\xa0Editor's Summary\\n\\xa0\\xa0\\xa0\\nThis copy is for your personal, non-commercial use only. \\nArticle Tools\\nhttp://science.sciencemag.org/content/early/2016/07/13/science.aaf9620tools: Visit the online version of this article to access the personalization and article\\nPermissions\\nhttp://www.sciencemag.org/about/permissions.dtlObtain information about reproducing this article: \\n is a registered trademark of AAAS. Science all rights reserved. The title Washington, DC 20005. Copyright 2016 by the American Association for the Advancement of Science;December, by the American Association for the Advancement of Science, 1200 New York Avenue NW, (print ISSN 0036-8075; online ISSN 1095-9203) is published weekly, except the last week inScience \\n on July 15, 2016 http://science.sciencemag.org/ Downloaded from \""
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def format_docs(docs):\n",
    "    return \"\\n\\n\".join(doc.page_content for doc in docs)\n",
    "pdf = '/home/v-haoguan/yuxuan/scihub.py/scihub/paper_HIV.pdf'\n",
    "loader = PyPDFLoader(\n",
    "file_path=pdf,\n",
    ")\n",
    "docs = loader.load()\n",
    "formatted_docs = format_docs(docs)\n",
    "formatted_docs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Response: Of course! Setting up a film club can be a lot of fun. Here are some steps and considerations to get you started:\n",
      "\n",
      "### 1. **Define Your Vision and Goals:**\n",
      "   - **Purpose:** Decide why you want to start the film club. Is it for entertainment, education, or a mix of both?\n",
      "   - **Audience:** Determine who your target audience is. Is it for students, professionals, a specific age group, etc.?\n",
      "\n",
      "### 2. **Gather Interested Individuals:**\n",
      "   - Spread the word among friends, colleagues, or through social media to find people who share your interest in films.\n",
      "   - Consider creating a simple online form to gauge interest and gather contact information.\n",
      "\n",
      "### 3. **Choose a Format:**\n",
      "   - Decide how often you will meet (e.g., weekly, bi-weekly).\n",
      "   - Determine the duration of each meeting (e.g., 1-2 hours).\n",
      "\n",
      "### 4. **Select Films:**\n",
      "   - Establish a method for selecting films (voting, rotating picks, themed months, etc.).\n",
      "   - Consider diversity in genres, languages, and eras to cater to different tastes.\n",
      "\n",
      "### 5. **Organize the Logistics:**\n",
      "   - Find a venue suitable for screenings (could be a member's home, community center, or a rented space).\n",
      "   - Arrange for necessary equipment (projector, screen, speakers, etc.).\n",
      "\n",
      "### 6. **Plan Discussions:**\n",
      "   - Decide how you will discuss each film (informal chat, structured discussion questions, guest speakers, etc.).\n",
      "   - Encourage respectful sharing of opinions and insights.\n",
      "\n",
      "### 7. **Promote Engagement:**\n",
      "   - Create a schedule of upcoming films and discussions.\n",
      "   - Encourage members to participate actively (e.g., by bringing snacks, suggesting films).\n",
      "\n",
      "### 8. **Consider Additional Activities:**\n",
      "   - Plan occasional outings to film festivals, screenings, or related events.\n",
      "   - Organize themed parties or film trivia nights to enhance camaraderie.\n",
      "\n",
      "### 9. **Manage Finances (if necessary):**\n",
      "   - If your club requires funds (for equipment rental, snacks, etc.), consider membership dues or seek sponsorship.\n",
      "\n",
      "### 10. **Evaluate and Adapt:**\n",
      "   - Regularly gather feedback from members to improve the club experience.\n",
      "   - Be open to adjusting the format or activities based on what works best for your group.\n",
      "\n",
      "By following these steps, you can create a vibrant film club that engages members and fosters a love for cinema. If you need more specific advice on any of these steps, feel free to ask!\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import getpass\n",
    "import requests\n",
    "\n",
    "# Prompt for API key securely\n",
    "api_key = getpass.getpass(prompt=\"Enter your OpenAI API Key: \")\n",
    "\n",
    "# OpenAI API endpoint\n",
    "url = \"https://api.wlai.vip/v1/chat/completions\"\n",
    "\n",
    "# Headers\n",
    "headers = {\n",
    "    \"Content-Type\": \"application/json\",\n",
    "    \"Authorization\": f\"Bearer {api_key}\"\n",
    "}\n",
    "\n",
    "# Request payload\n",
    "payload = {\n",
    "    \"model\": \"gpt-3.5-turbo\",  # or \"gpt-3.5-turbo\" if you're using GPT-3.5\n",
    "    \"messages\": [\n",
    "        {\"role\": \"system\", \"content\": \"You are a helpful assistant.\"},\n",
    "        {\"role\": \"user\", \"content\": \"Hello, can you help me set up a film club?\"}\n",
    "    ],\n",
    "    \"max_tokens\": 100\n",
    "}\n",
    "\n",
    "# Make the request\n",
    "response = requests.post(url, headers=headers, json=payload)\n",
    "\n",
    "if response.status_code == 200:\n",
    "    response_data = response.json()\n",
    "    print(\"Response:\", response_data['choices'][0]['message']['content'])\n",
    "else:\n",
    "    print(\"Failed to get a response:\", response.status_code, response.text)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sci-hub",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
